tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc

```diff
@@ -224,6 +227,280 @@ static bool HasCanonicalDimensionNumbers(
+class EinsumToLinalgConverter : public OpConversionPattern<mhlo::EinsumOp> {
+ public:
+  using OpConversionPattern<mhlo::EinsumOp>::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      mhlo::EinsumOp op, ArrayRef<Value> args,
+      ConversionPatternRewriter& rewriter) const final {
+    // Find maximum rank / number of loops.
+    auto get_rank = [](Value v) {
+      return v.getType().cast<ShapedType>().getRank();
+    };
+    auto is_scalar = [&](Value v) { return get_rank(v) == 0; };
+    auto einsum_config = op.einsum_config().str();
+
+    // Setting Up Commonly Used Symbols as strings
+    std::string arrow = "->";
+    std::string comma = ",";
+    std::string ellipsis = "...";
+
+    // With the assumption of binary Input Operand and Single Output
+    // Get the inputs and output operands' indices
+    // einsum_config = "lhs_loop,rhs_loop->output_loop"
+    std::size_t pos_arrow = einsum_config.find(arrow);
+    std::size_t pos_comma = einsum_config.find(comma);
+
+    std::string lhs_loop = einsum_config.substr(0, pos_comma);
+    std::string rhs_loop = einsum_config.substr(
+        pos_comma + comma.size(), pos_arrow - (pos_comma + comma.size()));
+    std::string output_loop = einsum_config.substr(pos_arrow + arrow.size());
+
+    // Check for Invalid Configs
+    // 1.Check that there is only maximum 2 inputs
+    // 2.Check that there is only maximum 1 output
+    // 3.Check that there is 1 arrow
+    if (rhs_loop.find(comma) != std::string::npos ||
+        output_loop.find(comma) != std::string::npos ||
+        output_loop.find(arrow) != std::string::npos) {
+      return rewriter.notifyMatchFailure(op, "Invalid Einsum Config!");
+    }
+
+    // Find result type, if on tensors.
+    Optional<ShapedType> result_ty;
+    result_ty = this->typeConverter->convertType(op->getResultTypes().front())
+                    .template dyn_cast<ShapedType>();
+
+    // Check result type compatibility.
+    if (!result_ty || !result_ty->hasRank() ||
+        !(result_ty->getElementType().isSignlessIntOrFloat() ||
+          result_ty->getElementType().isa<ComplexType>())) {
+      return rewriter.notifyMatchFailure(
+          op, "mismatched operand/result types or iterator count");
+    }
+
+    // Convert the representation from string/vector<char> to vector<string>
+    // i.e ("abc") -> {"a", "b", "c"} S.T we can have no limit on number of
+    // loops Since ASCII Chars has only 256 number of characters Find elilipsis
+    // in loop and replace and find number of dim/rank ellipsis represent then
+    // add batch dims into loop_dim as loop_dim.push_back({"0",... "N"}) where N
+    // is rank of batch then we can add the remaining dimensions. For example
+    // given config "ab...cde" representing where batch rank is 3 then we will
+    // get loop_dim = f("ab...cde") = {"a","b","0","1","2","c","d","e"}
+    std::vector<std::string> all_loops = {lhs_loop, rhs_loop, output_loop};
```

Comment:
Should be able to make it a `SmallVector<StringRef>` hopefully.

Reply:
done